linera-views/src/dynamo_db.rs

```diff
@@ -887,9 +823,208 @@ impl DynamoDbClientInternal {
+        match result {
+            None => {
+                if self.pos == self.iters.len() - 1 {
+                    return None;
+                }
+                self.pos += 1;
+                self.iters[self.pos]
+                    .next()
+                    .map(|x| extract_key(self.prefix_len, x))
+            }
+            Some(result) => Some(extract_key(self.prefix_len, result)),
+        }
+    }
+}
+
+/// A set of keys returned by a search query on DynamoDB.
+pub struct DynamoDbKeys {
+    result_queries: QueryResults,
+}
+
+impl KeyIterable<DynamoDbContextError> for DynamoDbKeys {
+    type Iterator<'a> = DynamoDbKeyBlockIterator<'a> where Self: 'a;
+
+    fn iterator(&self) -> Self::Iterator<'_> {
+        let pos = 0;
+        let mut iters = Vec::new();
+        for response in &self.result_queries.responses {
+            let iter = response.items.iter().flatten();
+            iters.push(iter);
+        }
+        DynamoDbKeyBlockIterator {
+            prefix_len: self.result_queries.prefix_len,
+            pos,
+            iters,
+        }
+    }
+}
+
+/// A set of `(key, value)` returned by a search query on DynamoDb.
+pub struct DynamoDbKeyValues {
+    result_queries: QueryResults,
+}
+
+#[doc(hidden)]
+#[allow(clippy::type_complexity)]
+pub struct DynamoDbKeyValueIterator<'a> {
+    prefix_len: usize,
+    pos: usize,
+    iters: Vec<
+        std::iter::Flatten<
+            std::option::Iter<'a, Vec<HashMap<std::string::String, AttributeValue>>>,
+        >,
+    >,
+}
+
+impl<'a> Iterator for DynamoDbKeyValueIterator<'a> {
+    type Item = Result<(&'a [u8], &'a [u8]), DynamoDbContextError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.iters[self.pos].next();
+        match result {
+            None => {
+                if self.pos == self.iters.len() - 1 {
+                    return None;
+                }
+                self.pos += 1;
+                self.iters[self.pos]
+                    .next()
+                    .map(|x| extract_key_value(self.prefix_len, x))
+            }
+            Some(result) => Some(extract_key_value(self.prefix_len, result)),
+        }
+    }
+}
+
+#[doc(hidden)]
+#[allow(clippy::type_complexity)]
+pub struct DynamoDbKeyValueIteratorOwned {
+    prefix_len: usize,
+    pos: usize,
+    iters: Vec<
+        std::iter::Flatten<
+            std::option::IntoIter<Vec<HashMap<std::string::String, AttributeValue>>>,
+        >,
+    >,
+}
+
+impl Iterator for DynamoDbKeyValueIteratorOwned {
+    type Item = Result<(Vec<u8>, Vec<u8>), DynamoDbContextError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.iters[self.pos].next();
+        match result {
+            None => {
+                if self.pos == self.iters.len() - 1 {
+                    return None;
+                }
+                self.pos += 1;
+                self.iters[self.pos]
+                    .next()
```

Comment:
(This will never be none, will it? None of the `iters` is empty from the beginning?)

Reply:
If it were to be empty then it would mean that the `DynamoDb` incorrectly set up a `LastEvaluatedKey` while there was noting else to do. I think we can safely exclude that scenario.